be2-scala/src/main/scala/ch/epfl/pop/storage/DbActorNew.scala

```diff
@@ -34,24 +34,88 @@ case class DbActorNew(
+      )
+    }
   }
 
   @throws [DbActorNAckException]
-  private def read(channel: Channel, messageId: Hash): Option[Message] = ???
+  private def read(channel: Channel, messageId: Hash): Option[Message] = {
+    Try(storage.read(s"$channel${Channel.DATA_SEPARATOR}$messageId")) match {
+      case Success(Some(json)) => Some(Message.buildFromJson(json))
+      case Success(None) => None
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def readChannelData(channel: Channel): ChannelData = ???
+  private def readChannelData(channel: Channel): ChannelData = {
+    Try(storage.read(channel.toString)) match {
+      case Success(Some(json)) => ChannelData.buildFromJson(json)
+      case Success(None) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, s"ChannelData for channel $channel not in the database")
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def readLaoData(channel: Channel): LaoData = ???
+  private def readLaoData(channel: Channel): LaoData = {
+    Try(storage.read(generateLaoDataKey(channel))) match {
+      case Success(Some(json)) => LaoData.buildFromJson(json)
+      case Success(None) => LaoData.emptyLaoData
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def catchupChannel(channel: Channel): List[Message] = ???
+  private def writeLaoData(channel: Channel, message: Message): Unit = {
+    this.synchronized{
+      val laoData: LaoData = Try(readLaoData(channel)) match {
+        case Success(data) => data
+        case Failure(_) => LaoData.emptyLaoData
+      }
+      val laoDataKey: String = generateLaoDataKey(channel)
+
+      Try(storage.write(
+        (laoDataKey, laoData.updateWith(message).toJsonString)
+      )).recover(
+        ex => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+      )
+    }
+  }
+
+  @throws [DbActorNAckException]
+  private def catchupChannel(channel: Channel): List[Message] = {
+
+    @scala.annotation.tailrec
+    def buildCatchupList(msgIds: List[Hash], acc: List[Message]): List[Message] = {
+      msgIds match {
+        case Nil => acc
+        case head :: tail =>
+          Try(read(channel, head)).recover(
+            ex => None
+          ) match {
+            case Success(Some(msg)) => buildCatchupList(tail, msg :: acc)
+            case _ => 
+              log.error(s"Problem encountered while catching up message with id $head")
+              buildCatchupList(tail, acc)
+          }
+      }
+    }
+
+    val channelData: ChannelData = Try(readChannelData(channel)) match {
+      case Success(data) => data
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
```

Comment:
You can simplify this (since `ex` is the same as what you are throwing at line 111)
```suggestion
    val channelData: ChannelData = readChannelData(channel)
```